test/openssl/test_x509req.rb

```diff
@@ -17,7 +17,11 @@ def issue_csr(ver, dn, key, digest)
     req = OpenSSL::X509::Request.new
     req.version = ver
     req.subject = dn
-    req.public_key = key.public_key
+    if key.oid == "ED25519"
+      req.public_key = key
+    else
+      req.public_key = key.public_key
+    end
```

Comment:
`key.public_key` is superfluous. We can use `req.public_key = key` with all pkey types, not just Ed25519.

Reply:
That method wasn't defined in my testing:

```
irb(main):001:0> require 'openssl'
=> true
irb(main):002:0> RUBY_VERSION
=> "3.2.5"
irb(main):003:0> OpenSSL::OPENSSL_VERSION
=> "OpenSSL 3.0.15 3 Sep 2024"
irb(main):004:0> ed25519 = OpenSSL::PKey::generate_key("ED25519")
=> #<OpenSSL::PKey::PKey:0x00007fa73112a9c8 oid=ED25519>
irb(main):005:0> ed25519.public_key
(irb):5:in `<main>': undefined method `public_key' for #<OpenSSL::PKey::PKey:0x00007fa73112a9c8 oid=ED25519> (NoMethodError)
```

Reply:
The `#public_key` method on PKey::RSA and PKey::DSA are not really useful and only kept for compatibility. https://ruby.github.io/openssl/OpenSSL/PKey/RSA.html#method-i-public_key

The `if` condition is not needed. Could you update it to not call `#public_key` on the pkey regardless of the key type?

Reply:
Sure I'll update that.

That said, passing the private `key` to the `csr.public_key=` setter is a little surprising to me. For example, the docs call `key.public_key` https://docs.ruby-lang.org/en/master/OpenSSL.html#module-OpenSSL-label-Certificate+Signing+Request

Reply:
@rhenium `test/openssl/test_x509req.rb` fails when I make your suggested change. It seems `req.public_key.to_der` is returning the much longer RSA private key, not the public key derived from the private one.

```
Failure: test_public_key(OpenSSL::TestX509Request)
/home/josh/work/ruby-openssl/test/openssl/test_x509req.rb:27:in `test_public_key'
     24: 
     25:   def test_public_key
     26:     req = issue_csr(0, @dn, @rsa1024, OpenSSL::Digest.new('SHA256'))
  => 27:     assert_equal(@rsa1024.public_key.to_der, req.public_key.to_der)
     28:     req = OpenSSL::X509::Request.new(req.to_der)
     29:     assert_equal(@rsa1024.public_key.to_der, req.public_key.to_der)
     30: 
<"0\x81\x9F0\r\x06\t*\x86H\x86\xF7\r\x01\x01\x01\x05\x00\x03\x81\x8D\x000\x81\x89\x02\x81\x81\x00\xCB\xC2\xC4\xB0\xD4@\xA7>\xD4\xFE>C\xA0\x1E\x17\x06\x03\xBDg\xC0-\xBF\x9C\xBF9T\x11\xA7F\xA0\xF1:\xA8\xD5\x87\xB0\xB1h\xA3\xC4E\x81\xEC\x93\x80O\n" +
"A7n\xBBS\x84\xF5\x9C\xF6H\xC7\x11\x04;\xB9\xFFX\xD6\xB6\xC2\xCFIZ\xC8\xDA\x87\xCB,\x10\x11R\xC5\x9A\x9D\\\xA4\x8B\x7FCx\x1E.\xFF\x19\x0F\xDAb\x86\x8C\n" +
"$<\x8C\x0E#z\x02\xB6\x14\x99\x973\xBDn=\xEF\xA3\x14\xDF\xE9y\xE0N\xA5\x17\xF2_\x14E9\x87\x02\x03\x01\x00\x01">
expected but was
<"0\x82\x02^\x02\x01\x00\x02\x81\x81\x00\xCB\xC2\xC4\xB0\xD4@\xA7>\xD4\xFE>C\xA0\x1E\x17\x06\x03\xBDg\xC0-\xBF\x9C\xBF9T\x11\xA7F\xA0\xF1:\xA8\xD5\x87\xB0\xB1h\xA3\xC4E\x81\xEC\x93\x80O\n" +
"A7n\xBBS\x84\xF5\x9C\xF6H\xC7\x11\x04;\xB9\xFFX\xD6\xB6\xC2\xCFIZ\xC8\xDA\x87\xCB,\x10\x11R\xC5\x9A\x9D\\\xA4\x8B\x7FCx\x1E.\xFF\x19\x0F\xDAb\x86\x8C\n" +
"$<\x8C\x0E#z\x02\xB6\x14\x99\x973\xBDn=\xEF\xA3\x14\xDF\xE9y\xE0N\xA5\x17\xF2_\x14E9\x87\x02\x03\x01\x00\x01\x02\x81\x81\x00\xA4\xA5\xFC\xC4\x1A\xAF'\xB5\xC8\xEC\xAC\xA9~\x8F\xF1\xF0\xC5B\xDE\xBCV\xFBW\x90\xD9\xA5J\"\x8E\x1A\xEC\xF2\x86\x8D\f\xF9ER\x15\xCB:\x93R\xF4\x99\xA0\xC4\xDB\x90\feH\x86\xCE\xE5\x87\xC9\xA59\xE5\xDE\xD1\\8\x1E\xB2@\xCCj@\xF5\xBF\xA6^\x8D\x8Bq&D\x97lD\xCEL\xB0\xF0Qd\xE7\xEB_\x13\x82v\xF5\xEE\xD55\x8EKT\xB5N\xBB\xC4\xAF\xF4\xF2/\xD7\xA1\xD9`\x9D\bl\xE9Z\aN\xE0\xA1\xDC{\xBB\xEC\xEE\x91\x02A\x00\xEF\xE81\xA8\e\x99G\x9E\xF6\xEF\x8F\xDA\x92%\xBFJ\xE5\xE6\xD1%\xCF\x12\xF8\xB2;S.I\x1C-#\xCE\xC1\xD3`\xF4\xAA\x05|\x0FU\aFn\x00\x84\xC9\xC9\xF0\xC0\xAC\xD7\xCD=\x90\xC4\x04&$\x14\n" +
"\xBA}\x99\x02A\x00\xD9m\xDAx\x01;^.\xB3\xA3g\x86\xE0,xP\xD3\a0\x87_\x05\x14\xBBV\n" +
"\x1E\x93\f\x9B\x1Ag\x1A\xA6\xA5th\x17\xB0\xE6A\xDEX\t\xB1?U\x94g\xF7`f\x1D\x90\xB4\xCDU\xBA @\xA9J\xA4\x1F\x02A\x00\xE2K\t\e\xE2\xFC\x90|7\xBC\xFC\xDAT\xE4\xDA-\xD1\xF2\b\xF2;\x03P&\xFE\xA2\x95\x94L\xC9\x9Fo\x15\x91GqA6\xEF0\x9F)\xBF\x9B_M\xE1\xF1c\xF0\xBA\x98\xCCu\xF8)\x8D(T\xFB\xBA\xAA\xF2q\x02A\x00\xC1\xAEl\x9C\xD7\xA4\x15\xCA\x8E4\xB2\x04\xE0n\\\xA2\xCA\xC8\xAD\xBE\xF8\xB2\xA2\xFC\x19\xB1\x9B\xF8\xAB\x93\x02\x9A\xF3\x8F\x9C\xF5+\xC0f\xD1E\xBD\x958\xD5\x0E0\xE2\xA9\x16 e\xF6R/\xECu?\xFEx\xFB\x14\xC5K\x02@cQq\x03\xF3w@\xF0\x97\n" +
"3\xE1\xE5\x8A\x85\xC4\x03\x10\xB3eUa7\xE0\xFA\xDD\f\x11\xC2Ct\xF3\x10\x0F\x92W\xEA\x0EG*q\xC5\x83i\x99\xB6\x85\xD0\xADi\x89J~\xE9\xF0bJ\xF3#S\xA2\x91\x04\xAC">
```

Reply:
Done